src/nni_manager/training_service/reusable/environments/remoteEnvironmentService.ts

```diff
@@ -0,0 +1,320 @@
+    private sshConnectionPromises: any[];
+    private experimentRootDir: string;
+    private experimentId: string;
+
+    constructor() {
+        super();
+        this.experimentId = getExperimentId();
+        this.environmentExecutorManagerMap = new Map<string, ExecutorManager>();
+        this.machineCopyExpCodeDirPromiseMap = new Map<RemoteMachineMeta, Promise<void>>();
+        this.machineExecutorManagerMap = new Map<RemoteMachineMeta, ExecutorManager>();
+        this.remoteMachineMetaOccupiedMap = new Map<RemoteMachineMeta, boolean>();
+        this.sshConnectionPromises = [];
+        this.experimentRootDir = getExperimentRootDir();
+        this.experimentId = getExperimentId();
+        this.log = getLogger();
+    }
+
+    public get prefetchedEnvironmentCount(): number {
+        return this.machineExecutorManagerMap.size;
+    }
+
+    public get environmentMaintenceLoopInterval(): number {
+        return 5000;
+    }
+
+    public get hasMoreEnvironments(): boolean {
+        return false;
+    }
+
+    public get hasStorageService(): boolean {
+        return false;
+    }
+
+    public async config(key: string, value: string): Promise<void> {
+        switch (key) {
+            case TrialConfigMetadataKey.MACHINE_LIST:
+                await this.setupConnections(value);
+                break;
+            case TrialConfigMetadataKey.TRIAL_CONFIG: {
+                const remoteMachineTrailConfig: TrialConfig = <TrialConfig>JSON.parse(value);
+                // Parse trial config failed, throw Error
+                if (remoteMachineTrailConfig === undefined) {
+                    throw new Error('trial config parsed failed');
+                }
+                // codeDir is not a valid directory, throw Error
+                if (!fs.lstatSync(remoteMachineTrailConfig.codeDir)
+                    .isDirectory()) {
+                    throw new Error(`codeDir ${remoteMachineTrailConfig.codeDir} is not a directory`);
+                }
+                try {
+                    // Validate to make sure codeDir doesn't have too many files
+                    await validateCodeDir(remoteMachineTrailConfig.codeDir);
+                } catch (error) {
+                    this.log.error(error);
+                    return Promise.reject(new Error(error));
+                }
+
+                this.trialConfig = remoteMachineTrailConfig;
+                break;
+            }
+            default:
+                this.log.debug(`Remote not support metadata key: '${key}', value: '${value}'`);
+        }
+    }
+
+    private scheduleMachine(): RemoteMachineMeta | undefined {
+        for (const [rmMeta, occupied] of this.remoteMachineMetaOccupiedMap) {
+            if (!occupied) {
+                this.remoteMachineMetaOccupiedMap.set(rmMeta, true);
+                return rmMeta;
+            }
+        }
+        return undefined;
+    }
+
+    private recycleMachineReservation(rmMeta: RemoteMachineMeta): void {
```

Comment:
it's called by one function only, and the name is confused with caller. don't need to be a seperated function.

Reply:
updated